semantic/monotype.go

```diff
@@ -442,20 +425,226 @@ func (l MonoType) Equal(r MonoType) bool {
+	}
 }
-func NewFunctionType() PolyType {
-	// TODO (algow): needs both a list of vars constraints and the monotype
-	return PolyType{}
+
+// monoTypeFromFunc will initialize a MonoType using the table
+// initialized from the function. If the property does not exist,
+// this will panic.
+func monoTypeFromFunc(fn func(obj *flatbuffers.Table) bool, t fbsemantic.MonoType) MonoType {
+	var table flatbuffers.Table
+	if !fn(&table) {
+		panic("table property missing")
+	}
+	mt, err := NewMonoType(table, t)
+	if err != nil {
+		panic(err)
+	}
+	return mt
 }
-func NewObjectType() MonoType {
-	// TODO (algow): needs both a list of vars constraints and the monotype
-	return MonoType{}
+
+// buildBasicType will construct a basic type in the builder
+// and return the offset for the type.
+func buildBasicType(builder *flatbuffers.Builder, t fbsemantic.Type) flatbuffers.UOffsetT {
+	fbsemantic.BasicStart(builder)
+	fbsemantic.BasicAddT(builder, t)
+	return fbsemantic.BasicEnd(builder)
+}
+
+// buildVarType will construct a var type in the builder
+// and return the offset for the type.
+func buildVarType(builder *flatbuffers.Builder, i uint64) flatbuffers.UOffsetT {
+	fbsemantic.VarStart(builder)
+	fbsemantic.VarAddI(builder, i)
+	return fbsemantic.VarEnd(builder)
+}
+
+// buildArrayType will construct an arr type in the builder
+// and return the offset for the type.
+func buildArrayType(builder *flatbuffers.Builder, elemType MonoType) flatbuffers.UOffsetT {
+	offset := copyMonoType(builder, elemType)
+	fbsemantic.ArrStart(builder)
+	fbsemantic.ArrAddTType(builder, elemType.mt)
+	fbsemantic.ArrAddT(builder, offset)
+	return fbsemantic.ArrEnd(builder)
+}
+
+// buildFunctionType will construct a fun type in the builder
+// and return the offset for the type.
+func buildFunctionType(builder *flatbuffers.Builder, retn MonoType, args []ArgumentType) flatbuffers.UOffsetT {
+	retnOffset := copyMonoType(builder, retn)
+	argsOffsets := make([]flatbuffers.UOffsetT, len(args))
+	for i, arg := range args {
+		nOffset := builder.CreateByteString(arg.Name)
+		tOffset := copyMonoType(builder, arg.Type)
+		fbsemantic.ArgumentStart(builder)
+		fbsemantic.ArgumentAddName(builder, nOffset)
+		fbsemantic.ArgumentAddTType(builder, arg.Type.mt)
+		fbsemantic.ArgumentAddT(builder, tOffset)
+		argsOffsets[i] = fbsemantic.ArgumentEnd(builder)
+	}
+
+	fbsemantic.FunStartArgsVector(builder, len(argsOffsets))
+	for i := len(argsOffsets) - 1; i >= 0; i-- {
+		builder.PrependUOffsetT(argsOffsets[i])
+	}
+	argsOffset := builder.EndVector(len(argsOffsets))
+
+	fbsemantic.FunStart(builder)
+	fbsemantic.FunAddRetnType(builder, retn.mt)
+	fbsemantic.FunAddRetn(builder, retnOffset)
+	fbsemantic.FunAddArgs(builder, argsOffset)
+	return fbsemantic.FunEnd(builder)
+}
+
+// buildObjectType will construct a row type in the builder
+// and return the offset for the type.
+func buildObjectType(builder *flatbuffers.Builder, properties []PropertyType) flatbuffers.UOffsetT {
```

Comment:
Shouldn't this take an optional type variable for records that are extended from one?